### DISCUSSION TITLE
cbc.exe Command Line Args to Library Calls

### DISCUSSION THREAD

**Discussion by cmsautter** [coin-or/Cbc#677] [Q&A] [answered]
Hello community,

I'm not very proficient in C++ and currently use Cbc as an executable in my Python application. Since Android prevents me from calling custom executables, I need to switch to the library version.

Could someone point me in the right direction on how to replicate the behavior of:
> cbc.exe in.mps -timeMode elapsed -branch -printingOptions all -solution out.sol

using library calls? I assume the main function is not part of the API?
Thanks a lot in advance for any help!

**Comment by jhmgoossens** [comment, marked as answer]
Have a look at the [example pool.cpp and search for CbcMain1](https://github.com/coin-or/Cbc/blob/d1dbc98911a224e2131d7e9db8596dad670430bc/examples/pool.cpp#L65). Does that help to get you started?

**Reply by cmsautter** [reply]
Yes, thank you for your quick reply! I now have an idea of how to proceed. I'll give you an update as soon as I find time to try it out!

**Reply by cmsautter** [reply]
After hours of fiddling, the solution is finally working in the Android app. For anyone wanting to replicate, or just for future reference, here are the steps performed, which couldn't be found online:

- First, create a small Python app using `ctypes` to access the functions `Cbc_getVersion, Cbc_newModel, Cbc_readMps, Cbc_solve, Cbc_bestSolution, Cbc_getNumCols`. These are sufficient to use `libCbcSolver.so` under WSL/Ubuntu. The 'Cbc_bestSolution' will map 1:1 to the list of lexicographically sorted variable names, so there's no need to generate the `.sol`.
- Note that the library from https://github.com/coin-or/python-mip/issues/392 will not run on Android, so it's necessary to build it manually.
- Refer to this explanation: https://www.reddit.com/r/termux/comments/xow8je/successful_compiling_coinor_cbc_solver_with/ for building instructions, but modify it slightly to use `c++14`:
```
# Set LDFLAGS, CFLAGS, and CXXFLAGS with the desired C++ standard
LDFLAGS="-L/data/data/com.termux/files/usr/lib" \
CFLAGS="-I/data/data/com.termux/files/usr/include/" \
CXXFLAGS="-std=c++14" \
./configure --build=aarch64-unknown-linux-gnu --prefix=$PREFIX/local
```
- Since Android doesn't support versioned `.so` files like `.so.1.2.3`, tweak the built libraries with `patchelf` to refer to the unversioned versions, which can then be bundled with the app.
- Lastly, also bundle `libbz2.so` and `libc++_shared.so` from the Termux environment with the app.

@jhmgoossens Thanks for pointing to `pool.cpp`; that was the minimal example needed to get started. Getting large C++ projects to build is the stuff of nightmares for me!

**Reply by tkralphs** [reply]
The instructions pointed to from Reddit say to get an old version of the source code that has all dependencies packaged, but I think this is bad advice. There have been many things fixed in the newer releases and getting the dependencies is easy with [coinbrew](https://github.com/coin-or/coinbrew).
